File: src/base/sort/bubble_sort.c
/**
 * 冒泡排序算法
 * 重复地走访过要排序的数列，一次比较两个元素，
 * 如果他们的顺序（如从大到小、首字母从A到Z）错误就把他们交换过来。
 * */
// Created by weiwei on 2019/1/27.
//
#include "sort.h"
#include <stdio.h>
int* bubble_sort(int * arr, int length) {
    int i, j, temp;
    for (i = 0; i < length - 1; i++)
        for (j = 0; j < length - 1 - i; j++)
            if (arr[j] > arr[j + 1]) {
                temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
            }
    return arr;
}

File: goalgrithm/sort/bubble_sort.go
package sort

// 冒泡排序
func BubbleSort(src []int) {
	size := len(src)
	for i := 0; i < size-1; i++ {
		for j := 0; j < size-i-1; j++ {
			if src[j] > src[j+1] {
				temp := src[j+1]
				src[j+1] = src[j]
				src[j] = temp
			}
		}
	}
}

// 改进版本冒泡排序

func BubbleSortOrderly(src []int) {
	orderly := false
	for i := 0; i < len(src)-1 && !orderly; i++ {
		orderly = true
		for j := 0; j < len(src)-1-i; j++ {
			if src[j] > src[j+1] {
				orderly = false
				temp := src[j+1]
				src[j+1] = src[j]
				src[j] = temp
			}
		}
	}
}
